jib-core/src/main/java/com/google/cloud/tools/jib/image/ImageLayers.java

```diff
@@ -19,34 +19,36 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
 import javax.annotation.Nullable;
 
 /** Holds the layers for an image. Makes sure that each layer is only added once. */
 public class ImageLayers<T extends Layer> implements Iterable<T> {
 
   public static class Builder<T extends Layer> {
 
-    private final ImmutableList.Builder<T> layersBuilder = ImmutableList.builder();
+    private final Set<T> layersBuilder = new LinkedHashSet<>();
     private final ImmutableSet.Builder<DescriptorDigest> layerDigestsBuilder =
         ImmutableSet.builder();
 
     /** The last layer added. */
     @Nullable private T lastLayer;
 
     /**
-     * Adds a layer.
+     * Adds a layer. Removes any prior occurrences of the same layer. Note that not all {@link
+     * Layer} types implement {@code equals/hashCode} - only those that are intended to be not
+     * duplicated will.
      *
      * @param layer the layer to add
      * @return this
      * @throws LayerPropertyNotFoundException if adding the layer fails
      */
     public Builder<T> add(T layer) throws LayerPropertyNotFoundException {
-      // Doesn't add the layer if the last layer is the same.
-      if (!isSameAsLastLayer(layer)) {
-        layerDigestsBuilder.add(layer.getBlobDescriptor().getDigest());
-        layersBuilder.add(layer);
-        lastLayer = layer;
-      }
+      layerDigestsBuilder.add(layer.getBlobDescriptor().getDigest());
+      layersBuilder.remove(layer);
+      layersBuilder.add(layer);
```

Comment:
I don't think `Set` allows duplicates, so probably there is no need to remove before adding.